src/main/java/mezz/jei/suffixtree/Node.java

```diff
@@ -113,7 +119,13 @@ private boolean contains(int index) {
 	}
 
 	void addEdge(char ch, Edge e) {
-		edges.put(ch, e);
+		try {
+			edges.put(ch, e);
+		} catch (UnsupportedOperationException ex) {
+			// after trimToSize() call - fall back
+			edges = new Char2ObjectOpenHashMap<>(edges);
+			edges.put(ch, e);
```

Comment:
Hitting this fallback a bunch could be pretty rough on performance, can you add a warning log here with the `ex` when it happens so we can trace down any mod that adds an ingredient super late?

Reply:
Added! Sorry that it took me a while. I'm not sure about what the warning should say to the end user exactly, though. (Note that this will run once per changed Node, not once per mod addition or once per any entry addition.)

(Also, this fallback is the reason why I moved the memory optimization process to the very end of the JEI loading lifecycle)